Chore: Aggregate save dashboard error

gastonqiu opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
To make the code more clearer, aggregate different save dashboard error
to DashboardErr struct. 

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes https://github.com/grafana/grafana/issues/15788

**Special notes for your reviewer**:

Commit:
Chore: Aggregate save dashboard error

To make the code more clear, aggregate different save dashboard error
to DashboardErr struct.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=26443) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=26443) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=26443) it.</sub>

Commit:
Chore: Use ShouldResemble instead of ShouldEqual

Use ShouldResemble instead of ShouldEqual. ShouldResemble does a
deep equal check. This can prevent the error raise because of type
difference.

aknuds1 commented:
Looking into this.

aknuds1 committed:
Use errors package for error detection

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Add Equal method to DashboardErr

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 reviewed (commented):
Generally looks like an improvement, but I see it can be simplified and I would also like it to use the standard [errors](https://golang.org/pkg/errors) package, to detect error types so we can have proper error hierarchies. Would you mind merging the changes from my [branch](https://github.com/grafana/grafana/tree/chore/dashboardErr) into yours?

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/dashboardErr

aknuds1 committed:
Document Equal

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

gastonqiu commented:
> Generally looks like an improvement, but I see it can be simplified and I would also like it to use the standard [errors](https://golang.org/pkg/errors) package, to detect error types so we can have proper error hierarchies. Would you mind merging the changes from my [branch](https://github.com/grafana/grafana/tree/chore/dashboardErr) into yours?
I see your code it's more better. Please do it.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/dashboardErr

aknuds1 reviewed (approved):
LGTM!

aknuds1 commented:
I pushed my changes directly to your branch, since you gave the OK. The current changes look good to me.

aknuds1 commented:
Adding @xlson since he opened the original issue.

papagian reviewed (changes requested):
Nice, can you rename `ErrDashboardUidToLong` -> `ErrDashboardUidTooLong` and its references as well?

aknuds1 commented on pkg/api/dashboard.go:
```diff
@@ -195,9 +196,15 @@ func deleteDashboard(c *models.ReqContext) Response {
+	if err != nil {
+		var dashboardErr models.DashboardErr
+		if ok := errors.As(err, &dashboardErr); ok {
+			// XXX: Should we allow other dashboard errors also?
```
@papagian I added that comment. Do you mind if we leave it? I added it since I'm unsure if it actually makes sense to 500 on all dashboard errors except for `ErrDashboardCannotDeleteProvisionedDashboard`.

papagian replied:
ok then

aknuds1 replied:
Actually, on second look, let's remove the comment. I see the connection now.

aknuds1 committed:
Update pkg/models/dashboards.go

Co-authored-by: Sofia Papagiannaki <papagian@users.noreply.github.com>

aknuds1 committed:
Merge remote-tracking branch 'gastonqiu/chore/dashboardErr' into chore/dashboardErr

aknuds1 committed:
Fix compiling

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 commented:
@papagian I committed your suggestion regarding renaming of an error and fixed references.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/dashboardErr

aknuds1 committed:
Fix tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into chore/dashboardErr

aknuds1 committed:
Update pkg/api/dashboard.go

Co-authored-by: Sofia Papagiannaki <papagian@users.noreply.github.com>

papagian commented on pkg/api/dashboard.go:
```diff
@@ -195,9 +196,14 @@ func deleteDashboard(c *models.ReqContext) Response {
-	} else if err != nil {
+	if err != nil {
+		var dashboardErr models.DashboardErr
+		if ok := errors.As(err, &dashboardErr); ok {
+			if errors.Is(err, models.ErrDashboardCannotDeleteProvisionedDashboard) {
```
On second look, I think that the first condition (`errors.As`) is redundant here. The 2nd one should be enough.

aknuds1 replied:
@papagian I thought the same for a second, but it's actually necessary to get the status code (`dashboardErr.StatusCode`).

papagian replied:
yes, I see now the problem.

xlson reviewed (approved):
This is a huge improvement. Thanks!

You could make a case for using an interface for the `errors.As(...)` checks as we aren't actually interested in the concrete types when doing that check, but it would add quite a bit of boilerplate and I don't see any big wins for a change like that at this point. When we need even more flexible errors we can easily make that change.

aknuds1 commented:
> You could make a case for using an interface for the `errors.As(...)` checks as we aren't actually interested in the concrete types when doing that check, but it would add quite a bit of boilerplate and I don't see any big wins for a change like that at this point. When we need even more flexible errors we can easily make that change.
@xlson Yeah I'm also unsure of how to _best_ design the error hierarchy, but agree we can iterate on this later :)

aknuds1 commented:
Thank you for contributing to Grafana!
